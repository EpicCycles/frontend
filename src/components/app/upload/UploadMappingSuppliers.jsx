import React, { Fragment } from 'react';
import { Button, Icon } from 'semantic-ui-react';
import * as PropTypes from 'prop-types';
import { removeKey } from '../../../helpers/utils';
import { getModelKey } from '../model/helpers/model';
import SupplierEdit from '../../supplier/SupplierEdit';

class UploadMappingSuppliers extends React.Component {
  constructor(props) {
    super();
    this.state = this.deriveStateFromProps(props);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }

  componentDidUpdate(prevProps) {
    // Typical usage (don't forget to compare props):
    if (this.props.suppliers !== prevProps.suppliers) {
      const updatedRowMappings = this.state.rowMappings.map(rowMap => {
        this.props.suppliers.some(supplier => {
          if (supplier.supplier_name.toLowerCase() === rowMap.supplierName.toLowerCase()) {
            rowMap.supplier = supplier.id;
            return true;
          }
          return false;
        });
        return rowMap;
      });
      this.setState({ rowMappings: updatedRowMappings });
    }
  }

  deriveStateFromProps = props => {
    const { rowMappings } = props;
    return {
      showModal: false,
      rowMappings,
    };
  };

  goToNextStep = () => {
    const { rowMappings } = this.state;
    this.props.addDataAndProceed({ rowMappings });
  };

  discardData = rowIndex => {
    const updatedRowMappings = this.state.rowMappings.map(rowMap => {
      if (rowMap.rowIndex === rowIndex) {
        let updatedRowMap = removeKey(rowMap, 'supplier');
        updatedRowMap.ignore = true;
        return updatedRowMap;
      }
      return rowMap;
    });
    this.setState({ rowMappings: updatedRowMappings });
  };

  undoDiscardData = rowIndex => {
    const updatedRowMappings = this.state.rowMappings.map(rowMap => {
      if (rowMap.rowIndex === rowIndex) {
        rowMap.ignore = false;
      }
      return rowMap;
    });
    this.setState({ rowMappings: updatedRowMappings });
  };

  setUpSupplierModalForNewField = rowMap => {
    const supplier = {
      supplier_name: rowMap.supplierName,
    };
    this.setState({
      supplier,
      showModal: true,
    });
  };

  handleCloseModal() {
    this.setState({ showModal: false, supplier: {} });
  }

  render() {
    const { saveSupplier, deleteSupplier } = this.props;
    const { rowMappings, showModal, supplier } = this.state;
    const unResolvedRowMappings = rowMappings.filter(
      rowMapping => !(rowMapping.supplier || rowMapping.ignore),
    );
    const discardedRowMappings = rowMappings.filter(rowMapping => rowMapping.ignore);
    return (
      <Fragment key="supplierProductUploadMapping">
        <section key="mappingData" className="row" id="mappingData">
          {showModal && (
            <SupplierEdit
              supplier={supplier ? supplier : {}}
              componentKey={getModelKey(supplier)}
              saveSupplier={saveSupplier}
              deleteSupplier={deleteSupplier}
              closeModal={this.handleCloseModal()}
            />
          )}
          {unResolvedRowMappings.map((mapping, index) => (
            <div key={`mapping${index}`} className="rounded">
              {mapping.supplierName}
              <Icon
                id={`delete-field${index}`}
                name="trash"
                onClick={() => this.discardData(mapping.rowIndex)}
                title="Discard data"
              />
              <Icon
                id={`create{index}`}
                name="add circle"
                onClick={() => this.setUpSupplierModalForNewField(mapping)}
                title="Create Supplier to store data"
              />
            </div>
          ))}
          {discardedRowMappings.map((mapping, index) => (
            <div key={`discard${index}`} className="rounded discarded">
              {mapping.supplierName}
              <Icon
                id={`restore-field${index}`}
                name="remove circle"
                onClick={() => this.undoDiscardData(mapping.rowIndex)}
                title="Do not Discard data"
              />
            </div>
          ))}
        </section>
        <div>
          <Button
            key="uploadCont"
            onClick={this.goToNextStep}
            disabled={unResolvedRowMappings.length > 0}
          >
            Continue ...
          </Button>
        </div>
      </Fragment>
    );
  }
}

UploadMappingSuppliers.propTypes = {
  rowMappings: PropTypes.array.isRequired,
  suppliers: PropTypes.array.isRequired,
  saveSupplier: PropTypes.func.isRequired,
  deleteSupplier: PropTypes.func.isRequired,
  addDataAndProceed: PropTypes.func.isRequired,
};
export default UploadMappingSuppliers;
