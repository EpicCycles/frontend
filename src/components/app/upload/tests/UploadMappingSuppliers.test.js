import React from 'react';
import toJson from 'enzyme-to-json';
import UploadMappingSuppliers from '../UploadMappingSuppliers';

const foundName = 'find me';
const suppliers = [
  { id: 1, supplier_name: 'id is 1' },
  { id: 2, supplier_name: foundName },
  { id: 3, supplier_name: 'id is 3' },
];
const rowMappings = [
  { rowIndex: 1, supplier: 2, supplierName: foundName },
  { rowIndex: 2, supplierName: 'id 11', ignore: true },
  { rowIndex: 3, supplierName: 'id 11' },
  { rowIndex: 4, supplierName: 'id 111' },
  { rowIndex: 5, supplierName: 'id 234' },
  { rowIndex: 11, supplier: 2, supplierName: foundName },
];
const saveSupplier = jest.fn();
const addDataAndProceed = jest.fn();

describe('UploadMappingSuppliers', () => {
  test('it renders', () => {
    const component = shallow(
      <UploadMappingSuppliers
        rowMappings={rowMappings}
        suppliers={suppliers}
        saveSupplier={saveSupplier}
        addDataAndProceed={addDataAndProceed}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  test('it sets up the supplierModal for a supplier', () => {
    const component = shallow(
      <UploadMappingSuppliers
        rowMappings={rowMappings}
        suppliers={suppliers}
        saveSupplier={saveSupplier}
        addDataAndProceed={addDataAndProceed}
      />,
    );
    component.instance().setUpSupplierModalForNewField({ rowIndex: 4, supplierName: 'id 111' });
    expect(component.state('showModal')).toBeTruthy();
    expect(component.state('supplier')).toEqual({ supplier_name: 'id 111' });
  });
  test('it marks a discarded row as ignore', () => {
    const component = shallow(
      <UploadMappingSuppliers
        rowMappings={rowMappings}
        suppliers={suppliers}
        saveSupplier={saveSupplier}
        addDataAndProceed={addDataAndProceed}
      />,
    );
    const rowMappingsExpected = [
      { rowIndex: 1, supplier: 2, supplierName: foundName },
      { rowIndex: 2, supplierName: 'id 11', ignore: true },
      { rowIndex: 3, supplierName: 'id 11' },
      { rowIndex: 4, supplierName: 'id 111' },
      { rowIndex: 5, supplierName: 'id 234' },
      { rowIndex: 11, supplierName: foundName, ignore: true },
    ];
    component.instance().discardData(11);
    expect(component.state('rowMappings')).toEqual(rowMappingsExpected);
  });
  test('it removes ignore when a row is removed from discard pile', () => {
    const component = shallow(
      <UploadMappingSuppliers
        rowMappings={rowMappings}
        suppliers={suppliers}
        saveSupplier={saveSupplier}
        addDataAndProceed={addDataAndProceed}
      />,
    );
    const rowMappingsExpected = [
      { rowIndex: 1, supplier: 2, supplierName: foundName },
      { rowIndex: 2, supplierName: 'id 11', ignore: false },
      { rowIndex: 3, supplierName: 'id 11' },
      { rowIndex: 4, supplierName: 'id 111' },
      { rowIndex: 5, supplierName: 'id 234' },
      { rowIndex: 11, supplier: 2, supplierName: foundName },
    ];

    component.instance().undoDiscardData(2);
    expect(component.state('rowMappings')).toEqual(rowMappingsExpected);
  });
  test('it calls passed method to save updated row mappings', () => {
    const component = shallow(
      <UploadMappingSuppliers
        rowMappings={rowMappings}
        suppliers={suppliers}
        saveSupplier={saveSupplier}
        addDataAndProceed={addDataAndProceed}
      />,
    );
    const rowMappingsExpected = [
      { rowIndex: 1, supplier: 2, supplierName: foundName },
      { rowIndex: 2, supplierName: 'id 11', ignore: false },
      { rowIndex: 3, supplierName: 'id 11' },
      { rowIndex: 4, supplierName: 'id 111' },
      { rowIndex: 5, supplierName: 'id 234' },
      { rowIndex: 11, supplier: 2, supplierName: foundName },
    ];

    component.instance().undoDiscardData(2);
    component.instance().goToNextStep();
    expect(addDataAndProceed.mock.calls).toHaveLength(1);
    expect(addDataAndProceed.mock.calls[0][0].rowMappings).toEqual(rowMappingsExpected);
  });
});
