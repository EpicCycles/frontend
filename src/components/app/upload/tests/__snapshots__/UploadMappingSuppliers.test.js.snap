// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadMappingSuppliers it renders 1`] = `
<Fragment
  key="supplierProductUploadMapping"
>
  <section
    className="row"
    id="mappingData"
    key="mappingData"
  >
    <div
      className="rounded"
      key="mapping0"
    >
      id 11
      <Icon
        as="i"
        id="delete-field0"
        name="trash"
        onClick={[Function]}
        title="Discard data"
      />
      <Icon
        as="i"
        id="create{index}"
        name="add circle"
        onClick={[Function]}
        title="Create Supplier to store data"
      />
    </div>
    <div
      className="rounded"
      key="mapping1"
    >
      id 111
      <Icon
        as="i"
        id="delete-field1"
        name="trash"
        onClick={[Function]}
        title="Discard data"
      />
      <Icon
        as="i"
        id="create{index}"
        name="add circle"
        onClick={[Function]}
        title="Create Supplier to store data"
      />
    </div>
    <div
      className="rounded"
      key="mapping2"
    >
      id 234
      <Icon
        as="i"
        id="delete-field2"
        name="trash"
        onClick={[Function]}
        title="Discard data"
      />
      <Icon
        as="i"
        id="create{index}"
        name="add circle"
        onClick={[Function]}
        title="Create Supplier to store data"
      />
    </div>
    <div
      className="rounded discarded"
      key="discard0"
    >
      id 11
      <Icon
        as="i"
        id="restore-field0"
        name="remove circle"
        onClick={[Function]}
        title="Do not Discard data"
      />
    </div>
  </section>
  <div>
    <Button
      as="button"
      disabled={true}
      key="uploadCont"
      onClick={[Function]}
    >
      Continue ...
    </Button>
  </div>
</Fragment>
`;
